tests/test_graph_utils.py

```diff
@@ -507,3 +509,103 @@ def test_result_equality(
+
+
+class TestSubgraphSampling:
+
+    @staticmethod
+    def get_graph(rng):
+        """
+        Randomly create somewhat complex graphs, then return a complex one
+        """
+        factory = GraphRandomFactory(
+            [0.15, 0.15, 0.15, 0.15, 0.14],
+            10, 0.15, rng)
+        # go through generation steps
+        for _ in range(5):
+            next(factory)  # type: ignore
+        return factory.graphs[0]
+
+    def test_sample(self):
+        """ Test sample respect given probability distribution """
+        pass
+
+    def test_uniform(self):
+        """ Sanity checks on uniform sampler """
+        rng = random.Random()
+        g = self.get_graph(rng)
+        n_vertices = max(1, len(g) - 4)
+        sg = uniform_sample(g, n_vertices, rng)
+        assert 1 <= len(sg) <= n_vertices
+        assert all(v in g for v in sg)
+
+    # Sometimes the RandomFactory fails (erode) or the algorithm does not converge
+    @pytest.mark.flaky(reruns=3)
+    def test_pagerank(self):
+        """ Sanity checks on pagerank sampler """
+        rng = random.Random()
+        g = self.get_graph(rng)
+        n_vertices = max(1, len(g) - 4)
+        sg = pagerank_sample(g, n_vertices, rng)
+        assert 1 <= len(sg) <= n_vertices
+
+    # Sometimes the RandomFactory fails (erode) or the algorithm does not converge
+    @pytest.mark.flaky(reruns=3)
+    def test_hubs(self):
+        """ Sanity checks on hubs sampler """
+        rng = random.Random()
+        g = self.get_graph(rng)
+        n_vertices = max(1, len(g) - 4)
+        sg = hubs_sample(g, n_vertices, rng)
+        assert 1 <= len(sg) <= n_vertices
+
+    # Sometimes the RandomFactory fails (erode) or the algorithm does not converge
+    @pytest.mark.flaky(reruns=3)
+    def test_authorities(self):
+        """ Sanity checks on authorities sampler """
+        rng = random.Random()
+        g = self.get_graph(rng)
+        n_vertices = max(1, len(g) - 4)
+        sg = authorities_sample(g, n_vertices, rng)
+        assert 1 <= len(sg) <= n_vertices
+
+    # Sometimes the RandomFactory fails (erode)
+    @pytest.mark.flaky(reruns=3)
+    def test_random_walk(self):
+        """ Sanity check on random walk sampler """
+        rng = random.Random()
+        g = self.get_graph(rng)
+        sg = random_walk_sample(g, rng, 10)
+        assert 1 <= len(sg) <= 10
+
+    # Sometimes the RandomFactory fails (erode)
+    @pytest.mark.flaky(reruns=3)
+    def test_random_specified_root(self):
+        """ Sanity check on random walk sampler """
+        rng = random.Random()
+        g = self.get_graph(rng)
+        seed = rng.choice(list(g))
+        sg = random_walk_sample(g, rng, 12, seeds=[seed])
+        assert 1 <= len(sg) <= 12
+        assert len(sg.over(seed)) <= 1 # empty, or self-reference
+
+    # Sometimes the RandomFactory fails (erode)
+    @pytest.mark.flaky(reruns=3)
+    def test_random_walk_multiple_roots(self):
```

Comment:
I had a failure when running the test (despite the reruns).
Maybe completely normal, but could you describe what's supposed to happen when it fails? (in relation to the comment).

In my case I got a failure on the assert:
assert 1 <= sum(1 for v in sg if len(sg.over(v)) <= 1) <= n_seeds

(the sum is equal to 5 and n_seed to 3)

Reply:
I'll look into it. As per the other non-related failures, I have the same issue with python 3.8 - and not with 3.7 :/

Reply:
Was indeed flaky. I was counting vertices which were the start of the sampled paths. Using `over` method, I thought to capture empty or self-ereference with the `len` only, which is wrong.

Setting up a stricter test solved the issue